rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md

```diff
@@ -0,0 +1,342 @@
+   will be assembled into the final object, and may be uploaded in any
+   order.
+3. The final object is assembled with the `CompleteMultipartUpload`
+   action.
+
+There are two major issues that this API presents that force some of
+the implementation decisions below:
+
+1. Vector may need to continue a multipart upload process across sink
+   restarts (due to reloads, restarts, or crashes). Since the uploads
+   require the presence of a unique identifier for both the multipart
+   upload itself as well as each of the uploaded parts, Vector will
+   need to store these identifiers with the partition that initiated
+   the upload.
+2. The upload parts (except for the final part) have a _minimum_ size
+   limit of 5MB. Since the sink may need to be restarted before a
+   batch reaches this minimum size, it will also need to buffer the
+   batch to disk before uploading the part.
+
+### User Experience
+
+The enhanced sink will add support for a new time-based upload
+mode. This mode will create a new object in the destination bucket
+once every configured interval instead of creating an object when a
+batch is filled or times out. This new object will be uploaded
+incrementally, but will not be visible in the bucket until all the
+parts of it are complete.
+
+The following upload behaviors are possible through the configuration
+described below. In all cases, where possible, the sink will upload
+the batch in multiple parts to reduce the amount of data buffered
+locally.
+
+1. The sink operates as before, where objects are created based only
+   on the existing batch settings. This will initially be the default
+   configuration to avoid breaking existing setups but will be
+   deprecated.
+1. The sink buffers batches until it can generate a complete part, and
+   assembles the parts into an object only at the end of an
+   interval. None of the intermediate parts are visible until the
+   final object is assembled.
+
+This change will introduce the following new configuration items for
+this sink, which will all default to operating the same as the
+existing sink.
+
+- `data_dir` (string): The path to the directory in which the state
+  data is stored. If this is not set, it is derived from the global
+  `data_dir` setting.
+- `batch.min_bytes` (integer): This specifies the target size a
+  locally-stored part must reach before it is uploaded. Due to the
+  limitations of the S3 API, this has an minimum value of 5MB, which
+  is also the default value when `upload_interval` is set.
+- `upload_interval` (string): When should the sink
+  complete the upload process and initiate a new one. If this is
+  unset, the existing batch-based object creation mechanism is
+  used. the following values are permitted:
+  - `"hourly"`: Complete the upload at the end of the hour exactly.
+  - `"daily"`: Complete the upload at midnight local time each day.
+  - `"at TIME"`: Complete the upload at a specific time each day.
```

Comment:
I'm not loving this option. A few reasons:

1. I think most users will want to base multipart finalization on an overall target size (i.e., 100mb) as opposed to a time range.
2. It's not clear if this time range is on the hour or within an hour of starting the multipart job.
3. It's not clear that this enables a multipart upload process that results in a final large object.

I think it may be better to provide settings that address the final object and follow the same pattern as `batch`. Call it `object`:

* `object.max_bytes` - Maximum number of bytes for the final object.
* `object.timeout_secs` - Object is finalized after this time regardless of the size.

What do you think?

Reply:
1. I was taking a cue from the [fluent s3 plugin](https://docs.fluentd.org/output/s3) which outputs objects hourly, albeit using local buffering instead of multipart uploads. It has a [timekey specifier](https://docs.fluentd.org/configuration/buffer-section#time) that controls this in a rather different way. If, however, an object size is the most desired configuration, I'll adjust for that.
2. Fair point. I had intended for `hourly` to mean on the hour.
3. The use of multipart vs single part should be transparent to the user. Essentially, the sink will upload parts _when it can_, and then create the object at the specified point (either time or size based).

I think having a specification for the final object size/time separate from the batch makes sense. I'm a little leery of the specific semantics of the settings described above, but the structure works.
1. Should the size limit be a hard maximum or a "close enough" maximum (ie if it goes one event over, that's OK)?
2. Is creating an object at the time of the first event + T (ie timeout-like behavior) more useful than fixed intervals (such as at exactly 00:00 every hour)?

Reply:
My preference would be to keep the batch configuration as-is, that is if I configure:

```toml
batch.max_bytes = 1073741824 # 1 GB
batch.max_events = 100_000
batch.timeout_secs = 9000
```

Then objects will be finalized from their parts when either the sum of the size of the parts would exceed the max, the sum of the number of events in the parts would exceed 100,000,  or 9000 seconds has passed.

We then add new configuration for the multi-part uploads like:

```toml
batch.min_part_bytes = 5242880 # 5 MB
```

Where Vector would attempt to build parts of at least this size to comprise the full batch.

If `min_part_bytes` exceeds `max_bytes` then we we do not do any partial uploads and the sink behaves as it does currently.

I think this would keep the batching behavior roughly in-line with other sinks.

I do agree with you, for the blob sinks, which typically include time specifiers in their partitioning it might make sense to augment them to be able to run "on the hour" or similar, but consider that outside of the scope of this RFC. It's something we could follow up on separately.

Reply:
Yes, this is a good simplification to reduce the scope of this effort. I will adjust accordingly.

Reply:
I'm ok with @jszwedko's proposal, but I do think it is slightly misleading. The finalization steps that reassembles the parts is entirely new and unique to S3. To me, it's completely separate from the batch+request process that exists today. Because of this, I think it deserves a new set of options. It's also worth thinking about doing this for other sink (i.e., the `file` sink) and what the options might look like.

Reply:
I agree with @binarylogic here about the configuration, but it leads to a bit of a conflict. If we add an `object.size` to control the desired size of the resulting object, how does that functionally differ from what `batch.max_bytes` does now? Should we deprecate and remove the latter?